### PR TITLE
Add ability to search apps from the MAS via name

### PR DESCRIPTION
**Pull request by maicki**
Addresses part of #9 by adding the ability to search for apps from the MAS via name. Currently the app id and the name will be printed.

### CODE REVIEW

**Comment by argon** [comment]
Please run `script/sort` within the repository and commit the changes

File: mas-cli/Commands/Search.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                   appId = result["trackId"] as? Int {
+                print("\(String(appId)) \(appName)")
+            }
+        }
+        
+        return .Success(())
+    }
+    
+    func searchURLString(appName: String) -> String {
+        return "https://itunes.apple.com/search?entity=macSoftware&term=\(appName)&attribute=allTrackTerm"
```
Comment (argon) [inline mas-cli/Commands/Search.swift]: This should probably url encode `appName` before inserting it into the URL.

File: mas-cli/NSURLSession+Synchronous.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+//  mas-cli
+//
+//  Created by Michael Schneider on 4/14/16.
+//  Copyright © 2016 Andrew Naylor. All rights reserved.
+//
+
+import Foundation
+
+/// NSURLSession synchronous behavior
+/// Particularly for playground sessions that need to run sequentially
```
Comment (argon) [inline mas-cli/NSURLSession+Synchronous.swift]: This file is from http://ericasadun.com/2015/11/12/more-bad-things-synchronous-nsurlsessions/. Please attribute correctly.

File: mas-cli/Commands/Search.swift [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+struct SearchCommand: CommandType {
+    typealias Options = SearchOptions
+    let verb = "search"
+    let function = "Search for apps from the Mac App Store"
+    
+    func run(options: Options) -> Result<(), MASError> {
+        let searchRequest = NSURLRequest(URL: NSURL(string: searchURLString(options.appName))!)
+        
+        guard let searchData = NSURLSession.requestSynchronousData(searchRequest),
+              let searchJsonString = try? NSJSONSerialization.JSONObjectWithData(searchData, options: []) as! Dictionary<String, AnyObject> else {
```
Comment (argon) [inline mas-cli/Commands/Search.swift]: Is there a reason not to use `NSURLSession.requestSynchronousJSONWithURLString` here?

**Comment by maicki** [comment]
@argon Addressed your comments let me know if there is more to do

File: mas-cli/Commands/Search.swift [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+struct SearchCommand: CommandType {
+    typealias Options = SearchOptions
+    let verb = "search"
+    let function = "Search for apps from the Mac App Store"
+    
+    func run(options: Options) -> Result<(), MASError> {
+        
+        guard let  searchURLString = searchURLString(options.appName),
+              let searchJson = NSURLSession.requestSynchronousJSONWithURLString(searchURLString) as? Dictionary<String, AnyObject> else {
```
Comment (argon) [inline mas-cli/Commands/Search.swift]: I would prefer to use the more compact form:
```swift
Dictionary<String, AnyObject> => [String: AnyObject]
```

File: mas-cli/Commands/Search.swift [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            return .Failure(MASError(code:.SearchError))
+        }
+        
+        guard let resultCount = searchJson["resultCount"] as? Int where resultCount > 0,
+              let results = searchJson["results"] as? Array<Dictionary<String, AnyObject>> else {
```
Comment (argon) [inline mas-cli/Commands/Search.swift]: As above:
```swift
[[String: AnyObject]]
```

File: mas-cli/Commands/Search.swift [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            print("No apps found")
```
Comment (argon) [inline mas-cli/Commands/Search.swift]: Could we please change this to "No results found" to be consistent with the error code naming below.

File: mas-cli/Commands/Search.swift [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            return .Failure(MASError(code:.NoSearchResultsFound))
+        }
+        
+        for result in results {
+            if let appName = result["trackName"] as? String,
+                   appId = result["trackId"] as? Int {
```
Comment (argon) [inline mas-cli/Commands/Search.swift]: Could we move all these key strings to static vars at the top to keep them in one place

```swift
struct ResultKeys {
    static let ResultCount = "resultCount"
    static let Results = "results"
    static let TrackName = "trackName"
    static let TrackId = "trackId"
}

File: mas-cli/Commands/Search.swift [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        
+        return .Success(())
+    }
+    
+    func searchURLString(appName: String) -> String? {
+        if let urlEncodedAppName = appName.URLEncodedString() {
+            return "https://itunes.apple.com/search?entity=macSoftware&term=\(urlEncodedAppName)&attribute=allTrackTerm"
+        }
+        return nil
+
```
Comment (argon) [inline mas-cli/Commands/Search.swift]: Unnecessary whitespace after return

File: mas-cli/main.swift [hunk 9]
```diff
@@ -17,6 +17,7 @@ public struct StderrOutputStream: OutputStreamType {
 let registry = CommandRegistry<MASError>()
 let helpCommand = HelpCommand(registry: registry)
 registry.register(AccountCommand())
+registry.register(SearchCommand())
```
Comment (argon) [inline mas-cli/main.swift]: I'm keeping the commands in alphabetical order for now, could you move `SearchCommand` down please.

**Comment by argon** [comment]
I made a few more comments. Sorry I didn't add these earlier, only had time for a quick look yesterday.

File: mas-cli/NSURLSession+Synchronous.swift [hunk 10]
```diff
@@ -0,0 +1,63 @@
+//
+//  NSURLSession+Synchronous.swift
+//  mas-cli
+//
+//  Created by Michael Schneider on 4/14/16.
+//  Copyright © 2016 Andrew Naylor. All rights reserved.
+//
+
+// Synchronouse NSURLSession code found at: http://ericasadun.com/2015/11/12/more-bad-things-synchronous-nsurlsessions/
```
Comment (argon) [inline mas-cli/NSURLSession+Synchronous.swift]: > Synchronous

no "e"

**Comment by maicki** [comment]
@argon Pushed new adjustments. Hope it's now a bit more to the style you prefer.

**Comment by argon** [comment]
Looks good to me, thanks for making those changes. :+1:
